presto-main/src/main/java/com/facebook/presto/cost/HistogramCalculator.java

```diff
@@ -0,0 +1,167 @@
+        Estimate min = histogram.inverseCumulativeProbability(0.0);
+        Estimate max = histogram.inverseCumulativeProbability(1.0);
+
+        // range is either above or below histogram
+        if ((!max.isUnknown() && max.getValue() < range.getLow())
+                || (!min.isUnknown() && min.getValue() > range.getHigh())) {
+            return Estimate.of(0.0);
+        }
+
+        // one of the max/min bounds can't be determined
+        if ((max.isUnknown() && !min.isUnknown()) || (!max.isUnknown() && min.isUnknown())) {
+            // when the range length is 0, the filter factor should be 1/distinct value count
+            if (!useHeuristics) {
+                return Estimate.unknown();
+            }
+
+            if (range.length() == 0.0) {
+                return totalDistinctValues.map(distinct -> 1.0 / distinct);
+            }
+
+            if (isFinite(range.length())) {
+                return Estimate.of(StatisticRange.INFINITE_TO_FINITE_RANGE_INTERSECT_OVERLAP_HEURISTIC_FACTOR);
+            }
+            return Estimate.of(StatisticRange.INFINITE_TO_INFINITE_RANGE_INTERSECT_OVERLAP_HEURISTIC_FACTOR);
+        }
+
+        // we know the bounds are both known, so calculate the percentile for each bound
+        // The inclusivity arguments can be derived from the open-ness of the interval we're
+        // calculating the filter factor for
+        // e.g. given a variable with values in [0, 10] to calculate the filter of
+        // [1, 9) (openness: false, true) we need the percentile from
+        // [0.0 to 1.0) (openness: false, true) and from [0.0, 9.0) (openness: false, true)
+        // thus for the "lowPercentile" calculation we should pass "false" to be non-inclusive
+        // (same as openness) however, on the high-end we want the inclusivity to be the opposite
+        // of the openness since if it's open, we _don't_ want to include the bound.
+        Estimate lowPercentile = histogram.cumulativeProbability(range.getLow(), openLow);
+        Estimate highPercentile = histogram.cumulativeProbability(range.getHigh(), !openHigh);
+
+        // both bounds are probably infinity, use the infinite-infinite heuristic
+        if (lowPercentile.isUnknown() || highPercentile.isUnknown()) {
+            if (!useHeuristics) {
+                return Estimate.unknown();
+            }
+            // in the case the histogram has no values
+            if (totalDistinctValues.equals(Estimate.zero()) || range.getDistinctValuesCount() == 0.0) {
+                return Estimate.of(0.0);
+            }
+
+            // in the case only one is unknown
+            if (((lowPercentile.isUnknown() && !highPercentile.isUnknown()) ||
+                    (!lowPercentile.isUnknown() && highPercentile.isUnknown())) &&
+                    isFinite(range.length())) {
+                return useHeuristics ? Estimate.of(StatisticRange.INFINITE_TO_FINITE_RANGE_INTERSECT_OVERLAP_HEURISTIC_FACTOR) : Estimate.unknown();
```

Comment:
Here should we directly return `Estimate.of(StatisticRange.INFINITE_TO_FINITE_RANGE_INTERSECT_OVERLAP_HEURISTIC_FACTOR)` as `useHeuristics` is always true?